dss/api/collections.py

```diff
@@ -46,43 +47,47 @@ def get_impl(uuid: str, replica: str, version: str = None):
-    return all_collections
 
 @dss_handler
 @security.authorized_group_required(['hca'])
 def listcollections(replica: str, per_page: int, start_at: int = 0):
+    """
+    Return a list of a user's collections.
+
+    Collection uuids are indexed and called by the user's email in a dynamoDB table.
+    Versions are collected by listing the files in the bucket with the prefix of that uuid.
+
+    :param Replica replica: AWS or GCP.
+    :param int per_page: # of collections returned per paged response.
+    :param int start_at: Where the next chunk of paged response should start at.
+    :return: A list of dictionaries looking like: [{'collection_uuid': uuid, 'collection_versions': [v1, v2]}, ... ].
+    """
+    owner = security.get_token_email(request.token_info)
+    uuids = get_collection_uuids_for_owner(owner)  # fetch from dynamodb index
     bucket = Replica[replica].bucket
     handle = Config.get_blobstore_handle(Replica[replica])
 
-    # expensively list every collection file in the bucket, even those not belonging to the user (possibly 1000's... )
-    collection_keys = [i for i in handle.list(bucket, prefix='collections') if not i.endswith('dead')]
+    collections = []
+    for uuid in uuids:
+        prefix = f'collections/{uuid}'
+        versions = [i.split('.', 1)[1] for i in handle.list(bucket, prefix=prefix) if not i.endswith('.dead')]
+        assert versions, f'{uuid} not found for user: {owner}'  # TODO: Determine how to handle this if it occurs.
+        collections.append({'collection_uuid': uuid,
+                            'collection_versions': versions})
 
     # paged response
-    if len(collection_keys) - start_at > per_page:
+    if len(collections) - start_at > per_page:
         next_url = UrlBuilder(request.url)
         next_url.replace_query("start_at", str(start_at + per_page))
-        # each chunk will be searched for collections belonging to that user (even more expensive; per bucket file)
-        # hits returned will vary between zero and the "per_page" size of the chunk
-        collections = fetch_collections(handle, bucket, collection_keys[start_at:start_at + per_page])
-        response = make_response(jsonify({'collections': collections}), requests.codes.partial)
+        collection_page = collections[start_at:start_at + per_page]
+        response = make_response(jsonify({'collections': collection_page}), requests.codes.partial)
         response.headers['Link'] = f"<{next_url}>; rel='next'"
         return response
     # single response returning all collections (or those remaining)
     else:
-        collections = fetch_collections(handle, bucket, collection_keys[start_at:])
-        return jsonify({'collections': collections}), requests.codes.ok
+        collection_page = collections[start_at:]
+        return jsonify({'collections': collection_page}), requests.codes.ok
+
 
 @dss_handler
 @security.authorized_group_required(['hca'])
```

Comment:
This `GET` will get updated for sharability at a later time?

Reply:
It will.  This PR got broken up into a number of smaller earlier PRs refactoring dynamodb in general, and the sharing functionality will also become a separate PR.